File: Circles.Profiles.Sdk/SafeLinkSigner.cs
using System.Numerics;
using Circles.Profiles.Interfaces;
using Circles.Profiles.Models.Core;
using Circles.Profiles.Sdk.Utils;
using Nethereum.ABI;
using Nethereum.Hex.HexConvertors.Extensions;
using Nethereum.Model;
using Nethereum.Signer;

namespace Circles.Profiles.Sdk;

/// <summary>
/// Produces an ECDSA signature with <c>signerAddress == SafeAddress</c>
/// while still using the *owner* EOA key for the cryptographic proof.
/// The resulting link passes on‑chain ERC‑1271 checks for the Safe.
/// </summary>
public sealed class SafeLinkSigner : ILinkSigner
{
    private readonly string _safe;
    private readonly IChainApi _chain;

    private static readonly ABIEncode Abi = new();

    // IMPORTANT: Safe expects "SafeMessage(bytes message)"
    private static readonly byte[] SafeMsgTypeHash =
        Sha3.Keccak256Bytes("SafeMessage(bytes message)"u8);

    private static readonly byte[] DomainTypeHash =
        Sha3.Keccak256Bytes("EIP712Domain(uint256 chainId,address verifyingContract)"u8);

    public SafeLinkSigner(string safeAddress, IChainApi chain)
    {
        bool isEmpty = string.IsNullOrWhiteSpace(safeAddress);
        if (isEmpty)
        {
            throw new ArgumentException(nameof(safeAddress));
        }

        _safe = safeAddress;
        _chain = chain;
    }

    public CustomDataLink Sign(CustomDataLink link, string ownerPrivKeyHex)
    {
        var ownerKey = new EthECKey(ownerPrivKeyHex);
        link = link with { SignerAddress = _safe };

        // keccak(canonical JSON w/o signature)
        byte[] payloadHash = Sha3.Keccak256Bytes(
            CanonicalJson.CanonicaliseWithoutSignature(link));

        // EIP‑712 Safe message hash for the *payload bytes* (via their keccak)
        byte[] safeHash = ComputeSafeHash(payloadHash, _chain.Id, _safe);

        var sig = ownerKey.SignAndCalculateV(safeHash);
        byte[] bytes = sig.To64ByteArray().Concat([sig.V[0]]).ToArray();

        return link with { Signature = "0x" + bytes.ToHex() };
    }

    public static byte[] BuildDomainSeparator(BigInteger chainId, string safe)
    {
        var encoded = Abi.GetABIEncoded(
            new ABIValue("bytes32", DomainTypeHash),
            new ABIValue("uint256", chainId),
            new ABIValue("address", safe));

        return Sha3.Keccak256Bytes(encoded);
    }

    /// <summary>
    /// Computes keccak(0x1901 || domainSeparator || keccak(SafeMessage(bytes message) || keccak(payloadBytes))).
    /// The <paramref name="payloadHash"/> is keccak(payloadBytes).
    /// </summary>
    public static byte[] ComputeSafeHash(byte[] payloadHash, BigInteger chainId, string safe)
    {
        // keccak256(SafeMessage(bytes message) || keccak(payloadBytes))
        byte[] safeMsg = Sha3.Keccak256Bytes(
            SafeMsgTypeHash.Concat(payloadHash).ToArray());

        // keccak256(0x1901 || domainSeparator || safeMsg)
        byte[] domain = BuildDomainSeparator(chainId, safe);
        return Sha3.Keccak256Bytes(new byte[] { 0x19, 0x01 }
            .Concat(domain)
            .Concat(safeMsg)
            .ToArray());
    }
}

File: js/src/SafeLinkSigner.ts
import { keccak256 } from 'js-sha3';
import { ethers, SigningKey } from 'ethers';
import { ILinkSigner } from './interfaces/ILinkSigner';
import { CustomDataLink } from './interfaces/CustomDataLink';
import { IChainApi } from './interfaces/IChainApi';
import { CanonicalJson } from './CanonicalJson';

/**
 * Produces an ECDSA signature with `signerAddress == SafeAddress`
 * while still using the *owner* EOA key for the cryptographic proof.
 * The resulting link passes on‑chain ERC‑1271 checks for the Safe.
 */
export class SafeLinkSigner implements ILinkSigner {
  private readonly safe: string;
  private readonly chain: IChainApi;
  private static readonly safeMsgTypeHash: Uint8Array = 
    Buffer.from(keccak256.create().update('SafeMessage(bytes)').digest());

  /**
   * Creates a new instance of SafeLinkSigner
   * @param safeAddress The address of the Safe wallet
   * @param chain The chain API
   */
  constructor(safeAddress: string, chain: IChainApi) {
    if (!safeAddress || safeAddress.trim() === '') {
      throw new Error('safeAddress is required');
    }

    this.safe = safeAddress;
    this.chain = chain;
  }

  /**
   * Signs a link with the provided owner private key
   * @param link The link to sign
   * @param ownerPrivKeyHex The owner's private key in hex
   * @returns The signed link
   */
  public sign(link: CustomDataLink, ownerPrivKeyHex: string): CustomDataLink {
    // Create a wallet from the owner private key
    const ownerWallet = new ethers.Wallet(ownerPrivKeyHex);
    
    // Update link with safe address as signer
    const updatedLink = {
      ...link,
      signerAddress: this.safe
    };

    // Calculate hashes
    const payloadHash = Buffer.from(
      keccak256.create().update(
        CanonicalJson.canonicaliseWithoutSignature(updatedLink)
      ).digest()
    );

    const safeTxHash = Buffer.from(
      keccak256.create().update(
        Buffer.concat([SafeLinkSigner.safeMsgTypeHash, payloadHash])
      ).digest()
    );

    const chainId = BigInt(this.chain.getChainId());
    const domainSeparator = this.buildDomainSeparator(chainId, this.safe);

    const safeHash = Buffer.from(
      keccak256.create().update(
        Buffer.concat([
          Buffer.from([0x19, 0x01]),
          domainSeparator,
          safeTxHash
        ])
      ).digest()
    );

    // Sign the hash
    const messageHashBytes = ethers.getBytes(safeHash);
    
    // Use SigningKey directly for better compatibility
    const signingKey = new SigningKey(ownerPrivKeyHex);
    const signature = signingKey.sign(messageHashBytes);
    
    // Create the 65-byte signature (r + s + v)
    const r = signature.r.substring(2); // remove 0x
    const s = signature.s.substring(2); // remove 0x
    const v = signature.v.toString(16).padStart(2, '0');
    
    const signatureHex = `0x${r}${s}${v}`;

    // Return the signed link
    return {
      ...updatedLink,
      signature: signatureHex
    };
  }

  /**
   * Builds the domain separator for EIP-712
   * @param chainId The chain ID
   * @param safeAddress The safe address
   * @returns The domain separator as a byte array
   */
  private buildDomainSeparator(chainId: bigint, safeAddress: string): Uint8Array {
    const typeHash = Buffer.from(
      keccak256.create().update('EIP712Domain(uint256 chainId,address verifyingContract)').digest()
    );

    // Encode parameters according to Solidity ABI encoding
    const abiCoder = ethers.AbiCoder.defaultAbiCoder();
    const encoded = abiCoder.encode(
      ['bytes32', 'uint256', 'address'],
      [typeHash, chainId, safeAddress]
    );

    return Buffer.from(keccak256.create().update(Buffer.from(encoded.substring(2), 'hex')).digest());
  }
}
